src/theory/quantifiers/ceg_t_instantiator.cpp

```diff
@@ -855,37 +889,122 @@ Node BvInverter::getInversionSkolemFor( Node cond, TypeNode tn ) {
+  TypeNode ftn = NodeManager::currentNM()->mkFunctionType( NodeManager::currentNM()->booleanType(), tn );
+  return getSolveVariable( ftn );
+}
+
+Node BvInverter::getInversionNode( Node cond, TypeNode tn ) {
+  // condition should be rewritten
+  Node new_cond = Rewriter::rewrite(cond);
+  if( new_cond!=cond ){
+    Trace("bv-invert-debug") << "Condition " << cond << " was rewritten to " << new_cond << std::endl;
+  } 
+  Node f = getInversionSkolemFunctionFor( tn );
+  return NodeManager::currentNM()->mkNode( kind::APPLY_UF, f, new_cond );
+}
+
+bool BvInverter::isInvertible( Kind k ) {
+  // TODO : make this precise (this should correspond to all kinds that we handle in solve_bv_lit/solve_bv_constraint)
+  return k!=APPLY_UF;
+}
+
 Node BvInverter::getPathToPv( Node lit, Node pv, Node sv, std::vector< unsigned >& path, std::unordered_set< TNode, TNodeHashFunction >& visited ) {
   if( visited.find( lit )==visited.end() ){
     visited.insert( lit );
     if( lit==pv ){
       return sv;
     }else{
-      unsigned rmod = 0; // TODO : randomize?
-      for( unsigned i=0; i<lit.getNumChildren(); i++ ){
-        unsigned ii = ( i + rmod )%lit.getNumChildren();
-        Node litc = getPathToPv( lit[ii], pv, sv, path, visited );
-        if( !litc.isNull() ){
-          // path is outermost term index last
-          path.push_back( ii );
-          std::vector< Node > children;
-          if( lit.getMetaKind() == kind::metakind::PARAMETERIZED ){
-            children.push_back( lit.getOperator() );
-          }
-          for( unsigned j=0; j<lit.getNumChildren(); j++ ){
-            children.push_back( j==ii ? litc : lit[j] );
+      // only recurse if the kind is invertible 
+      // this allows us to avoid paths that go through skolem functions
+      if( isInvertible( lit.getKind() ) ){
+        unsigned rmod = 0; // TODO : randomize?
+        for( unsigned i=0; i<lit.getNumChildren(); i++ ){
+          unsigned ii = ( i + rmod )%lit.getNumChildren();
+          Node litc = getPathToPv( lit[ii], pv, sv, path, visited );
+          if( !litc.isNull() ){
+            // path is outermost term index last
+            path.push_back( ii );
+            std::vector< Node > children;
+            if( lit.getMetaKind() == kind::metakind::PARAMETERIZED ){
+              children.push_back( lit.getOperator() );
+            }
+            for( unsigned j=0; j<lit.getNumChildren(); j++ ){
+              children.push_back( j==ii ? litc : lit[j] );
+            }
+            return NodeManager::currentNM()->mkNode( lit.getKind(), children );
           }
-          return NodeManager::currentNM()->mkNode( lit.getKind(), children );
         }
       }
     }
   }
   return Node::null();
 }
 
+Node BvInverter::eliminateSkolemFunctions( TNode n, std::vector< Node >& side_conditions, std::unordered_map< TNode, Node, TNodeHashFunction >& visited ) {
+  std::unordered_map< TNode, Node, TNodeHashFunction >::iterator it = visited.find( n );
```

Comment:
I would prefer to not make this recursive. We are dealing with bit-vector terms here that can be deeply nested. It would also make this method redundant and the traversing part can be moved to `eliminateSkolemFunctions`  in line [999](https://github.com/CVC4/CVC4/pull/1164/files#diff-b5ab4785f7e81f4f2bebb8629bcfb16cR999).